### PR TITLE
Retry Execute when reserved connection is lost

### PR DESCRIPTION
**Pull request by systay**
When using long lived connections and reserved connections, one can easily run into the query timeout limit. This leads to queries failing temporarily, but succeeding if when query is retried, since the connection has been cleaned up and can be re-established.

This PR does the retry once automatically at the VTGate level.

Fixes: https://github.com/vitessio/vitess/issues/6862

### CODE REVIEW

File: go/test/endtoend/vtgate/reservedconn/sysvar_test.go [hunk 1]
```diff
@@ -183,14 +183,9 @@ func TestSetSystemVarWithConnectionTimeout(t *testing.T) {
...
-	// first query will fail
+	// connection has timed out, but vtgate will recreate the connection for us
 	_, err = exec(t, conn, `select @@sql_safe_updates from test where id = 80`)
-	require.Error(t, err)
-
-	// this query is able to succeed.
-	newQR := checkedExec(t, conn, `select @@sql_safe_updates, connection_id() from test where id = 80`)
-	require.Equal(t, qr.Rows[0][0], newQR.Rows[0][0])
-	require.NotEqual(t, qr.Rows[0][1], newQR.Rows[0][1])
+	require.NoError(t, err)
```
Comment (harshit-gangal) [inline go/test/endtoend/vtgate/reservedconn/sysvar_test.go:188]: either use `checkedExec` or `assertMatches`

File: go/test/endtoend/vtgate/reservedconn/sysvar_test.go [hunk 2]
```diff
@@ -286,9 +281,9 @@ func TestSetSystemVarAutocommitWithConnError(t *testing.T) {
 	// first query to 80- shard should pass
 	assertMatches(t, conn, "select id, val1 from test where id = 4", "[[INT64(4) NULL]]")
 
-	// first query to -80 shard will fail
+	// first query to -80 shard will fail, but vtgate should retry once and succeed the second time
 	_, err = exec(t, conn, "insert into test (id, val1) values (2, null)")
-	require.Error(t, err)
+	require.NoError(t, err)
```
Comment (harshit-gangal) [inline go/test/endtoend/vtgate/reservedconn/sysvar_test.go:286]: same as above

File: go/test/endtoend/vtgate/reservedconn/sysvar_test.go [hunk 3]
```diff
@@ -324,9 +319,9 @@ func TestSetSystemVarInTxWithConnError(t *testing.T) {
 	checkedExec(t, conn, "rollback")
 	assertMatches(t, conn, "select id, val1 from test where id = 2", "[]")
 
-	// first query to 80- shard will fail
+	// first query to -80 shard will fail, but vtgate should retry once and succeed the second time
 	_, err = exec(t, conn, "select @@sql_safe_updates from test where id = 4")
-	require.Error(t, err)
+	require.NoError(t, err)
```
Comment (harshit-gangal) [inline go/test/endtoend/vtgate/reservedconn/sysvar_test.go:324]: same as above

File: go/vt/vtgate/scatter_conn.go [hunk 4]
```diff
@@ -220,8 +220,15 @@ func (stc *ScatterConn) ExecuteMultiShard(
 			case nothing:
 				innerqr, err = qs.Execute(ctx, rs.Target, queries[i].Sql, queries[i].BindVariables, info.transactionID, info.reservedID, opts)
 				if err != nil {
-					checkAndResetShardSession(info, err, session)
-					return nil, err
+					shouldRetry := checkAndResetShardSession(info, err, session)
+					if shouldRetry {
+						// we seem to have lost our connection. if it was a reserved connection, let's try to recreate it
+						info.actionNeeded = reserve
+						innerqr, reservedID, alias, err = qs.ReserveExecute(ctx, rs.Target, session.SetPreQueries(), queries[i].Sql, queries[i].BindVariables, info.transactionID, opts)
```
Comment (harshit-gangal) [inline go/vt/vtgate/scatter_conn.go:227]: info.transactionID can be replaced with `0` as this case is only allowed for non-transactional connection.

File: go/vt/vttablet/sandboxconn/sandboxconn.go [hunk 5]
```diff
@@ -127,8 +128,10 @@ func (sbc *SandboxConn) getError() error {
 		sbc.MustFailCodes[code] = count - 1
 		return vterrors.New(code, fmt.Sprintf("%v error", code))
 	}
-	if sbc.ShardErr != nil {
-		return sbc.ShardErr
+	if sbc.EphemeralShardErr != nil {
+		err := sbc.EphemeralShardErr
+		sbc.EphemeralShardErr = nil
+		return err
```
Comment (harshit-gangal) [inline go/vt/vttablet/sandboxconn/sandboxconn.go:134]: could use `sbc.MustFailCodes` with count 1 in place of EphemeralShardErr

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vttablet/sandboxconn/sandboxconn.go (hunk 5 above)
Comment (systay) [reply go/vt/vttablet/sandboxconn/sandboxconn.go:134]: not really - I need a specific `SQLError` here, and `MustFailCodes` only supports `vterrors`, right?

**Pull request merged**